sympy/core/expr.py

```diff
@@ -40,26 +40,67 @@ def _diff_wrt(self):
         temporarily converts the non-Symbol vars in Symbols when performing
         the differentiation.
 
-        Note, see the docstring of Derivative for how this should work
-        mathematically. In particular, note that expr.subs(yourclass, Symbol)
-        should be well-defined on a structural level, or this will lead to
-        inconsistent results.
+        Non-symbol _diff_wrt=True variables must also implement a
+        _eval_derivative_wrt method. See Expr._eval_derivative_wrt
+        for details.
 
         Examples
         ========
 
-        >>> from sympy import Expr
+        >>> from sympy import Expr, Dummy
         >>> e = Expr()
         >>> e._diff_wrt
         False
         >>> class MyClass(Expr):
         ...     _diff_wrt = True
         ...
+        ...     def _eval_derivative_wrt(self, expr, new_name):
+        ...         new_self = Dummy(new_name)
+        ...         new_expr = expr.subs(self, new_self)
+        ...         return (new_expr, new_self)
+        ...
         >>> (2*MyClass()).diff(MyClass())
         2
         """
         return False
 
+    def _eval_derivative_wrt(self, expr, new_name):
```

Comment:
From the comments on https://github.com/sympy/sympy/pull/7437, I don't think `new_name` is a needed argument here. The function can just use a `Dummy`, and pick its name to be whatever it wants (the name of a dummy doesn't matter, except for printing).